crates/aptos/src/account/key_rotation.rs

```diff
@@ -94,56 +118,131 @@ impl CliCommand<RotateSummary> for RotateKey {
+        let (new_private_key, new_public_key) = if new_derivation_path.is_some() {
+            (
+                None,
+                aptos_ledger::get_public_key(new_derivation_path.clone().unwrap().as_str(), false)?,
+            )
+        } else {
+            let new_private_key = self
+                .extract_private_key(self.txn_options.encoding_options.encoding)?
+                .ok_or_else(|| {
+                    CliError::CommandArgumentError("Unable to parse new private key".to_string())
+                })?;
+            (Some(new_private_key.clone()), new_private_key.public_key())
+        };
+
+        // Check that public key is actually changing.
+        if new_public_key == current_public_key {
             return Err(CliError::CommandArgumentError(
-                "New private key cannot be the same as the current private key".to_string(),
+                "New public key cannot be the same as the current public key".to_string(),
             ));
         }
 
-        // Get sequence number for account
-        let sequence_number = self.txn_options.sequence_number(sender_address).await?;
-        let auth_key = self.txn_options.auth_key(sender_address).await?;
-
+        // Construct rotation proof challenge.
+        let sequence_number = self.txn_options.sequence_number(current_address).await?;
+        let auth_key = self.txn_options.auth_key(current_address).await?;
         let rotation_proof = RotationProofChallenge {
             account_address: CORE_CODE_ADDRESS,
             module_name: "account".to_string(),
             struct_name: "RotationProofChallenge".to_string(),
             sequence_number,
-            originator: sender_address,
+            originator: current_address,
             current_auth_key: AccountAddress::from_bytes(auth_key)
                 .map_err(|err| CliError::UnableToParse("auth_key", err.to_string()))?,
-            new_public_key: new_private_key.public_key().to_bytes().to_vec(),
+            new_public_key: new_public_key.to_bytes().to_vec(),
         };
-
         let rotation_msg =
             bcs::to_bytes(&rotation_proof).map_err(|err| CliError::BCS("rotation_proof", err))?;
 
-        // Signs the struct using both the current private key and the next private key
-        let rotation_proof_signed_by_current_private_key =
-            current_private_key.sign_arbitrary_message(&rotation_msg.clone());
-        let rotation_proof_signed_by_new_private_key =
-            new_private_key.sign_arbitrary_message(&rotation_msg);
+        // Sign the struct using both the current private key and the new private key.
+        let rotation_proof_signed_by_current_private_key = if current_derivation_path.is_some() {
+            eprintln!("Approve rotation proof challenge signature on your Ledger device");
+            aptos_ledger::sign_message(
+                current_derivation_path.clone().unwrap().as_str(),
```

Comment:
Nit: `if let Some` (or personally I prefer just good old `match`) to avoid the `is_some` -> `.unwrap()` pattern.

Reply:
Addressed in 5b9dc70